### PR TITLE
[RFR] Export breakpoints

### PR DESCRIPTION
**Pull request by HugoGiraudel**

### CODE REVIEW

File: plugins/im-export-breakpoints/README.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+.nav {
+  @include media('>=medium') {
+    // Do the styling here
+  }
+}
+```
+
+```javascript
+function doCalculations() {
```
Comment (HugoGiraudel) [inline plugins/im-export-breakpoints/README.md]: I don't think a README should be a blog post, i.e. I don't think it should contain unrelated code snippets.

File: plugins/im-export-breakpoints/README.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  // Do calculations here
+}
+```
+
+This works fine, but we're declaring the `768px` breakpoint twice. Not DRY!
+
+### With this plugin
+
+```scss
+$breakpoints: (small: 320px, medium: 768px, large: 1024px);
```
Comment (HugoGiraudel) [inline plugins/im-export-breakpoints/README.md]: Aren't we supposed to quote strings?

File: plugins/im-export-breakpoints/README.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+This plugin grabs all the breakpoints from **include-media** and outputs their status as JSON format on the `content` property of a DOM object (`<body>` by default), allowing developers to make decisions based on the viewport width without having to re-declare their breakpoints, leading to maintainability problems.
+
+## Example
+
+Let's say that you need to do calculations to create a grid layout, but doing them for mobile viewports is a waste because they will just go full width anyway.
+
+### Without this plugin
+
+```scss
+$breakpoints: (small: 320px, medium: 768px, large: 1024px);
```
Comment (HugoGiraudel) [inline plugins/im-export-breakpoints/README.md]: Aren't we supposed to quote strings?

File: plugins/im-export-breakpoints/README.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+.nav {
+  @include media('>=medium') {
+    // Do the styling here
+  }
+}
+```
+
+```javascript
+function doCalculations() {
```
Comment (HugoGiraudel) [inline plugins/im-export-breakpoints/README.md]: I don't think a README should be a blog post, i.e. I don't think it should contain unrelated code snippets.

File: plugins/im-export-breakpoints/includeMedia.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+  function setElement(newElement) {
+    element = newElement;
+  }
+
+  function readBreakpoints() {
+    if (window.getComputedStyle && (window.getComputedStyle(element).content != '')) {
+      var breakpoints = window.getComputedStyle(element).content.replace(/'/g, '');
+
+      return JSON.parse(breakpoints);
```
Comment (HugoGiraudel) [inline plugins/im-export-breakpoints/includeMedia.js]: Maybe `try`/`catch` in case JSON is invalid?

File: plugins/im-export-breakpoints/_include-media-export-breakpoints.scss [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    $state: $target-value >= $value;
+    $json: $json + '"#{$name}":#{$state}';
+  }
+
+  $json: $json + '}';
+
+  @return $json;
+}
+
+@mixin im-export-breakpoints($element: body) {
```
Comment (HugoGiraudel) [inline plugins/im-export-breakpoints/_include-media-export-breakpoints.scss]: Where is the mixin called?

Comment (eduardoboucas) [inline plugins/im-export-breakpoints/_include-media-export-breakpoints.scss]: Anywhere really. Do you think it should be called automatically since you're actively including the plugin file?

File: plugins/im-export-breakpoints/_include-media-export-breakpoints.scss [hunk 7]
```diff
@@ -0,0 +1,24 @@
+@function im-breakpoints-to-json($target-name) {
```
Comment (HugoGiraudel) [inline plugins/im-export-breakpoints/_include-media-export-breakpoints.scss]: Could (should?) be rewritten like this:

```scss
@function im-breakpoints-to-json($target-name) {
  $json: ();
  $target-value: map-get($breakpoints, $target-name);

  @each $name, $value in $breakpoints {
    $json: append($json, '"#{$name}": #{$target-value >= $value}', 'comma');
  }

  @return '{#{$json}}' ;
}
```

File: plugins/im-export-breakpoints/_include-media-export-breakpoints.scss [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+  $json: $json + '}';
+
+  @return $json;
+}
+
+@mixin im-export-breakpoints($element: body) {
+  @each $name, $value in $breakpoints {
+    @include media('>=#{$name}') {
+      #{$element} {
```
Comment (HugoGiraudel) [inline plugins/im-export-breakpoints/_include-media-export-breakpoints.scss]: We should probably use a (hidden) pseudo-element rather than a direct element?

File: plugins/im-export-breakpoints/_include-media-export-breakpoints.scss (hunk 6 above)
Comment (HugoGiraudel) [inline plugins/im-export-breakpoints/_include-media-export-breakpoints.scss]: Might be interesting?

Comment (HugoGiraudel) [inline plugins/im-export-breakpoints/_include-media-export-breakpoints.scss]: Also you did not document it.

File: plugins/im-export-breakpoints/_include-media-export-breakpoints.scss (hunk 8 above)
Comment (eduardoboucas) [inline plugins/im-export-breakpoints/_include-media-export-breakpoints.scss]: I think in general `<body>` is a good option because it's faster to query. Also, [`support for getComputedStyle()`] on pseudo-elements is slightly more restrictive than on normal elements. But it's worth thinking about.

File: plugins/im-export-breakpoints/_include-media-export-breakpoints.scss (hunk 6 above)
Comment (eduardoboucas) [inline plugins/im-export-breakpoints/_include-media-export-breakpoints.scss]: I'm cool with calling it automatically.

What kind of documentation would you add for the JS file?

File: plugins/im-export-breakpoints/_include-media-export-breakpoints.scss (hunk 8 above)
Comment (HugoGiraudel) [inline plugins/im-export-breakpoints/_include-media-export-breakpoints.scss]: Is there any problem with using `content` on body? Wouldn't it display the content in some browsers?

File: plugins/im-export-breakpoints/README.md (hunk 4 above)
Comment (eduardoboucas) [inline plugins/im-export-breakpoints/README.md]: Agree. Will write a proper article over the weekend.

File: plugins/im-export-breakpoints/_include-media-export-breakpoints.scss (hunk 6 above)
Comment (HugoGiraudel) [inline plugins/im-export-breakpoints/_include-media-export-breakpoints.scss]: Where to include it basically. There is no mention of the JS file whatsoever in the documentation.

Comment (eduardoboucas) [inline plugins/im-export-breakpoints/_include-media-export-breakpoints.scss]: > Import includeMedia.js onto your project. This file is just a simplistic approach to access the information sent across by include-media. Feel free to extend it to fit your needs.

I think it's a bit difficult to add more information on how to include the file without making assumptions about the JS architecture being used. What are your thoughts?

File: plugins/im-export-breakpoints/_include-media-export-breakpoints.scss (hunk 7 above)
Comment (eduardoboucas) [inline plugins/im-export-breakpoints/_include-media-export-breakpoints.scss]: Definitely. Will change it.

File: plugins/im-export-breakpoints/_include-media-export-breakpoints.scss (hunk 6 above)
Comment (HugoGiraudel) [inline plugins/im-export-breakpoints/_include-media-export-breakpoints.scss]: True that.

File: plugins/im-export-breakpoints/includeMedia.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
+var includeMedia = (function () {
```
Comment (HugoGiraudel) [inline plugins/im-export-breakpoints/includeMedia.js]: Shall we name it `im`?

File: plugins/im-export-breakpoints/README.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+```
+
+We're grabbing the value of `medium` from **include-media**, so if we need to update it we'll just do it in the Sass.
+
+## Installation
+
+### Sass
+
+```scss
+@import '_include-media-export-breakpoints';
```
Comment (HugoGiraudel) [inline plugins/im-export-breakpoints/README.md]: Remove leading underscore. Also, let's name it `im-export`, because this file name is longer than my arm right now.

File: plugins/im-export-breakpoints/_include-media-export-breakpoints.scss (hunk 8 above)
Comment (eduardoboucas) [inline plugins/im-export-breakpoints/_include-media-export-breakpoints.scss]: Potentially. Pseudo-element is probably safer.

File: plugins/im-export-breakpoints/includeMedia.js (hunk 9 above)
Comment (eduardoboucas) [inline plugins/im-export-breakpoints/includeMedia.js]: Yeah!

File: plugins/im-export-breakpoints/README.md (hunk 10 above)
Comment (eduardoboucas) [inline plugins/im-export-breakpoints/README.md]: But we main dist file has a leading underscore. Won't that be a bit inconsistent?

I'm cool with `im-export` anyways.

File: plugins/im-export-breakpoints/includeMedia.js (hunk 5 above)
Comment (eduardoboucas) [inline plugins/im-export-breakpoints/includeMedia.js]: :ok_hand:

File: plugins/im-export-breakpoints/README.md (hunk 2 above)
Comment (eduardoboucas) [inline plugins/im-export-breakpoints/README.md]: Will remove these snippets anyway.

File: plugins/im-export-breakpoints/README.md (hunk 3 above)
Comment (eduardoboucas) [inline plugins/im-export-breakpoints/README.md]: Will remove these snippets anyway.

File: plugins/im-export-breakpoints/README.md (hunk 10 above)
Comment (HugoGiraudel) [inline plugins/im-export-breakpoints/README.md]: Keep the underscore in the file name but not in the import.

**Comment by eduardoboucas** [comment]
@HugoGiraudel changed the following:

- Renamed mixin to `im-export`;
- Mixin runs automatically;
- Element is always an `::after`. By default, it's on `<body>`, but it's possible to override that by declaring the variable `$im-export-element` with the desired element's selector;
- Added information about the current active breakpoint to the JSON and created the function `getActive()` in the JS;
- Renamed the JS object to `im`;
- Wrapped the JSON parsing in a `try/catch` (even though I'm not doing anything with the error at the moment — thoughts?).

File: plugins/im-export/_include-media-export.scss [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+  $target-value: map-get($breakpoints, $target-name);
+
+  @each $name, $value in $breakpoints {
+    $json: append($json, '"#{$name}":#{$target-value >= $value}', 'comma');
+  }
+
+  @return '{#{$json}}' ;
+}
+
+@mixin im-export($element) {
```
Comment (HugoGiraudel) [inline plugins/im-export/_include-media-export.scss]: Add SassDoc.

File: plugins/im-export/_include-media-export.scss [hunk 12]
```diff
@@ -0,0 +1,23 @@
+@function im-breakpoints-to-json($target-name) {
```
Comment (HugoGiraudel) [inline plugins/im-export/_include-media-export.scss]: Add SassDoc.

File: plugins/im-export/includeMedia.js [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+  }
+
+  function getActiveBreakpoint() {
+    var breakpoints = readBreakpoints();
+    
+    if (!breakpoints) {
+      return false;
+    }
+
+    return breakpoints['_'];
```
Comment (HugoGiraudel) [inline plugins/im-export/includeMedia.js]: What is this?

File: plugins/im-export/_include-media-export.scss [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+  $json: ('"_": "#{$target-name}"');
```
Comment (HugoGiraudel) [inline plugins/im-export/_include-media-export.scss]: What is this?

File: plugins/im-export/_include-media-export.scss [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+/// Generates the media queries necessary to export breakpoints
+///
+/// @param {String} $element - Element to append JSON data to
+///
+@mixin im-export($element) {
+  @each $name, $value in $breakpoints {
+    @include media('>=#{$name}') {
+      #{$element} {
+        content: im-breakpoints-to-json($name);
+        display: none;
```
Comment (HugoGiraudel) [inline plugins/im-export/_include-media-export.scss]: Nope. See: https://github.com/HugoGiraudel/SassyJSON/blob/master/stylesheets/encode/mixins/_json.scss#L18.

File: plugins/im-export/includeMedia.js [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+  function readBreakpoints() {
+    var result = false;
+
+    if (window.getComputedStyle && (window.getComputedStyle(element, '::after').content != '')) {
+      var imData = window.getComputedStyle(element, '::after').content.replace(/'/g, '');
+      var result = false;
+
+      try {
+        result = JSON.parse(imData);
+      } catch(err) {}
```
Comment (HugoGiraudel) [inline plugins/im-export/includeMedia.js]: Log an error?

**Comment by eduardoboucas** [comment]
@HugoGiraudel I've restructured things a bit, so this is how it looks like now:

- Each breakpoint is a node inside the main JSON object. The property `value` is always present, and the property `active` is added (and set to true) if the current breakpoint is greater than or equal to the breakpoint in question.

*Example (for a current viewport width of 900px):*
```json
{
    "phone": {
        "value": "320px",
        "active": true
    },
    "tablet": {
        "value": "768px",
        "active": true
    },
    "desktop": {
        "value": "1024px"
    }
}
```

- On the JavaScript side, the API has 5 methods:
  - `setElement()`: changes the default element (`body`) to any element (optional)
  - `greaterThan(breakpoint-name)`: returns whether the current viewport is greater than or equal to the given breakpoint — it should technically be called `greaterThanOrEqualTo()`, but it's a bit verbose, what do you think?
  - `lessThan(breakpoint-name)`: opposite of `greaterThan()`
  - `getLargestActive()`: returns the largest breakpoint active. For the example above, would return `"tablet"`. Again, not a big fan of the function name, but couldn't think of anything better. Suggestions?
  - `getValue(breakpoint-name, as-number)`: returns the value for a given breakpoint, with the option to return either a string (including the units) or a number — for the example above, would return `"768px"` or `768` respectively.

**Comment by eduardoboucas** [comment]
Updated the README as well.

**Comment by HugoGiraudel** [comment]
> The property value is always present, and the property active is added (and set to true) if the current breakpoint is greater than or equal to the breakpoint in question.

It should exist in all case. Having a consistent data structure is much better than performing some blind checking. Make it always live, `true` or `false`.

> it should technically be called greaterThanOrEqualTo(), but it's a bit verbose, what do you think?

Yep. Let's keep it like this.

> Added the ability to choose between manual or automatic updates. By default it will read the DOM element and parse the JSON every time any function is called, but with setUpdateMode('manual') it will read once and cache the result until update() is called.

Isn't it over-engineering?

**Comment by eduardoboucas** [comment]
> It should exist in all case. Having a consistent data structure is much better than performing some blind checking. Make it always live, true or false.

Okay, sounds good to me.

> Isn't it over-engineering?

It's on the verge of becoming too complex, yes, but I think it makes sense. If you want to perform scarce checks on the active breakpoints, you'll want the library to update automatically so you can run one command. However, if you need to heavily use it you may require to take control over the process.

I think the best example is if you need to attach checks to the scrolling event. It becomes too expensive to do that every time the user scrolls, but if the `scroll` event reads from the cached version and the `resize` events updates that cache, performance is much better.

I agree that this is as far as we should go with the library, but I think it's a perfectly reasonable use case for this and I'd rather implement it than forcing people to write their own stuff on top of it.

**Comment by eduardoboucas** [comment]
Changed the structure to always have `active`. 

Let me know if you're okay with me merging this.

File: plugins/include-media-export/README.md [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+
+This plugin grabs all the breakpoints from **include-media** and outputs their status as JSON format on the `content` property of a DOM object (`<body>` by default), allowing developers to make decisions based on the viewport width without having to re-declare their breakpoints, leading to maintainability problems.
+
+More information about this plugin can be found [here](https://eduardoboucas.com/blog/2016/06/29/using-breakpoints-in-javascript-with-include-media.html).
+
+## Installation
+
+### Sass
+
+Download [this file](#) and include it in your Sass project
```
Comment (HugoGiraudel) [inline plugins/include-media-export/README.md]: Missing link.

File: plugins/include-media-export/README.md [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+
+```scss
+@import 'include-media-export';
+```
+
+### JavaScript
+
+Import `includeMedia.js` onto your project. This file is just a simplistic approach to access the information sent across by include-media. Feel free to extend it to fit your needs.
```
Comment (HugoGiraudel) [inline plugins/include-media-export/README.md]: Do we camelCase JS filenames?

File: plugins/include-media-export/README.md [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+
+## API
+
+###`im.greaterThan(breakpoint)`
```
Comment (HugoGiraudel) [inline plugins/include-media-export/README.md]: Missing space after `###`.

File: plugins/include-media-export/README.md [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+  - `breakpoint` - Name of the breakpoint to test against
+- **Returns:** `Boolean`
+- **Example:**
+```javascript
+if (im.greaterThan('desktop')) {
+    console.log('This is a really big screen');
+}
+```
+---
+###`im.lessThan(breakpoint)`
```
Comment (HugoGiraudel) [inline plugins/include-media-export/README.md]: Missing space after `###`.

File: plugins/include-media-export/README.md [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+- **Accepts:** N/A
+- **Returns:** `String`
+- **Example:**
+```javascript
+if (im.getActive() == 'phone') {
+    console.log('This looks like a phone');
+}
+```
+---
+###`im.getValue(breakpoint, asNumber)`
```
Comment (HugoGiraudel) [inline plugins/include-media-export/README.md]: Missing space after `###`.

File: plugins/include-media-export/README.md [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+- **Accepts:**
+  - `breakpoint` - Name of the breakpoint
+  - `asNumber` (optional) - Return the value as a `String` (with units) or as a number (unitless) 
+- **Returns:** `String` or `Float`
+- **Example:**
+```javascript
+console.log('For me, a desktop has a width of ' + im.getValue('desktop'));
+```
+---
+###`im.setElement(element)`
```
Comment (HugoGiraudel) [inline plugins/include-media-export/README.md]: Missing space after `###`.

File: plugins/include-media-export/README.md [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+
+- **Accepts:**
+  - `element` - DOM element
+- **Returns:** N/A
+- **Example:**
+```javascript
+im.setElement(document.getElementById('my-element'));
+```
+---
+###`im.setUpdateMode(mode)`
```
Comment (HugoGiraudel) [inline plugins/include-media-export/README.md]: Missing space after `###`.

File: plugins/include-media-export/README.md [hunk 24]
```diff
@@ -0,0 +1,113 @@
...
+
+- **Accepts:**
+  - `mode` - Update mode. `auto` (default) or `manual`
+- **Returns:** N/A
+- **Example:**
+```javascript
+im.setUpdateMode('manual');
+```
+---
+###`im.update()`
```
Comment (HugoGiraudel) [inline plugins/include-media-export/README.md]: Missing space after `###`.

File: plugins/include-media-export/_include-media-export.scss [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+///
+/// @example json - For a width of 900px
+///  '{"phone":{"value": "320px", "active": true}, "tablet":{"value": "768px", "active": true}, "desktop":{"value": "1024px", "active": false}}'
+///
+@function im-breakpoints-to-json($target-name) {
+  $breakpoints-json: ();
+  $target-value: map-get($breakpoints, $target-name);
+
+  @each $name, $value in $breakpoints {
+    $breakpoint: '"#{$name}":{"value": "#{$value}", "active": #{$target-value >= $value}}';
```
Comment (HugoGiraudel) [inline plugins/include-media-export/_include-media-export.scss]: Missing space after `: `.

File: plugins/include-media-export/includeMedia.js [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+  var element = document.body;
+  var autoUpdate = true;
+  var breakpoints = false;
+
+  function setElement(newElement) {
+    element = newElement;
+  }
+
+  function setUpdateMode(updateMode) {
+    if (updateMode == 'manual') {
```
Comment (HugoGiraudel) [inline plugins/include-media-export/includeMedia.js]: Loosy equal.

File: plugins/include-media-export/includeMedia.js [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+      } catch(err) {}
+    }
+  }
+
+  function isBreakpointActive(breakpoint) {
+    if (autoUpdate) {
+      readBreakpoints();
+    }
+
+    return breakpoints.hasOwnProperty(breakpoint) &&
```
Comment (HugoGiraudel) [inline plugins/include-media-export/includeMedia.js]: Any reason to go that specific with `breakpoints[breakpoint].hasOwnProperty('active') && breakpoints[breakpoint].active` rather than `breakpoints[breakpoint].active`?

File: plugins/include-media-export/includeMedia.js [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+           breakpoints[breakpoint].hasOwnProperty('active') &&
```
Comment (HugoGiraudel) [inline plugins/include-media-export/includeMedia.js]: Any reason to go that specific with `breakpoints[breakpoint].hasOwnProperty('active') && breakpoints[breakpoint].active` rather than `breakpoints[breakpoint].active`?

File: plugins/include-media-export/includeMedia.js [hunk 29]
```diff
@@ -0,0 +1,82 @@
...
+  function getBreakpointValue(breakpoint, asNumber) {
+    if (autoUpdate) {
+      readBreakpoints();
+    }
+
+    if (!breakpoints || !breakpoints.hasOwnProperty(breakpoint)) {
+      return false;
+    }
+
+    return asNumber ? parseFloat(breakpoints[breakpoint].value) : breakpoints[breakpoint].value;
```
Comment (HugoGiraudel) [inline plugins/include-media-export/includeMedia.js]: Idea: instead of `asNumber` relying on `parseFloat`, accept a function that is being applied on the value.

```javascript
if (typeof fn === 'function') {
  return fn(breakpoints[breakpoint].value);
}
```

```javascript
var value = im.getBreakpointValue('small', parseFloat);
```

**Comment by HugoGiraudel** [comment]
Added a few comments. :)

**Comment by eduardoboucas** [comment]
I'll close this because the current version of the plugin is [here](https://github.com/eduardoboucas/include-media-export). I made note of all yours comments though and will make sure to rectify the issues. :+1:
